### ISSUE TITLE
MAAT Restart Bugfixing and Add Ons

### ISSUE THREAD

**Issue by abbeyjohnson1** [walkeranthonyp/MAAT#29]
**Crossover Probability Bug in Restart Process**

- Currently, the selection of crossover probabilities is being recalculated with each restart, which we do not want to happen. This works fine for the initial MCMC simulation, but it crashes the code on the first restart (in most instances). 
- This can be fixed by "carrying over" the adapted `p_CR` vector and other relevant variables between restarts (and continuing to adapt them across restarts if necessary). 
- We also need to rethink the size of `CR_burnin`. Currently, it is being calculated as a user-inputted percentage of `mcmc_maxiter`. For most MCMC simulations we would want a crossover burn-in size of around 10,000.
- The following crossover variables are the ones that need to be carried over between restarts: `p_CR`, `L` (index of which crossover probabilities are selected), `del` (squared normalized jumping distance), and `m` (index of indexes).

**Output Files Bug**

- The output log file is being overwritten with each restart.
- The "history" output file is being overwritten with each restart as well.
- Ideally, it would be nice to have each history file and corresponding log file; this would allow for more comprehensive problem solving and post-processing.

**Refining the Outputting and Convergence-testing Processes**
- We would like to put a switch in place that holds off on testing for convergence until after 10,000 MCMC samples.
- We would also like an additional switch that "flips" after convergence has been reached. It would perhaps be ideal to store all samples post-convergence separate from the other samples.

**Additional Thoughts**

- It could be informative to users if we carried over the "original j" value (i.e., total number of samples since the beginning of the MCMC process). It would be nice to output this value in the log files in addition to the history and output RDS files. If many restarts are required, this could help the user keep track of where they are in the MCMC process.
- In the most recent version of the code, all of the variables that need to be passed through between restarts will be in a `mcmc_pars` list, a sublist of `wpars`.

**Comment by abbeyjohnson1** [comment]
**Update**
_Correction to the last bullet point in the "Additional Thoughts" section:_
Not all of the variables that need to be passed through between restarts are are in the `wpars$mcmc_pars` sublist. And not all of the of the variables in `wpars$mcmc_pars` need to be shared between restarts. The lists are structured the way that they are in an attempt to achieve naming consistency, but this may make things difficult to keep track of what needs to be passed through between restarts. 

 The `mcmc_pars` sublist includes:
- `start_iter`
- `run_type` 
- `lklihood` 
- `outlier` 
- `conv` 
- `bdry_handling`
- `init_prior` 
- `chains` 
- `maxiter`
- `start_iter`
- `thin`   
- `thin_obs`  
- `homosced`
- `chain_delta` 
- `c_rand`        
- `c_ergod`  
- `p_gamma`   
- `n_CR`   
- `adapt_pCR`   
- `CR_burnin` 
- `check_ss`  
- `check_iter`

These variables are in this list, because they are designed to be able to be quickly changed by the user.

The `mcmc` list includes:
- `outlier_detected`
- `j_start_burnin`
- ` j_burnin50` 
- `d`   
- `CR`          
- `p_CR`          
- `R`          
- `current_state`   
- `p_state`        
- `sd_state`    
- `jump`            
- `draw`         
- `lambda`   
- `boundary_min`  
- `boundary_max` 
- `del`    
- `L`     
- `t`          
- `m`           
- `CR_burnin`     
- `d_star`     

These variables are in this list, because they are "internal" algorithmic variables that don't need to be touched by the user. 

The variables that need to be passed through between restarts are listed here:
-  original`j` (discussed above)
- `n_CR`   
- `adapt_pCR`   
- `CR_burnin` 
- `CR`          
- `p_CR`
- `del`    
- `L`     
- `t`          
- `m`

**Comment by walkeranthonyp** [comment]
Thanks for the comprehensive description of this issue @abbeyjohnson1 . I'm planning to pick this up today and over the next week, so there might be more comments here or on slack. One very quick question I have now: What do you mean by the "history" file? Is it the MCMC history?

**Comment by abbeyjohnson1** [comment]
It's my pleasure @walkeranthonyp! Thank you for working on this, and please feel free to reach out to me at any time!

So a "history" file is an output created after each run and restart run. The way we designed it was so that there would be one primary output file that contains the information we want during post-processing (the last 50% of all total samples, corresponding model evals, etc...). This is one file that is appended with each restart run. The given "history" file for each run contains all of the output we wouldn't need unless we were troubleshooting or doing a more detailed post-processing analysis. There is a new "history" file created with each restart run. But if we put all the history files together, they would contain the first 50% of all total samples and corresponding outputs.

Please let me know if this doesn't answer your question!

**Comment by abbeyjohnson1** [comment]
It's my pleasure @walkeranthonyp! Thank you for working on this, and please feel free to reach out to me at any time!

So a "history" file is an output created after each run and restart run. The way we designed it was so that there would be one primary output file that contains the information we want during post-processing (the last 50% of all total samples, corresponding model evals, etc...). This is one file that is appended with each restart run. The given "history" file for each run contains all of the output we wouldn't need unless we were troubleshooting or doing a more detailed post-processing analysis. There is a new "history" file created with each restart run. But if we put all the history files together, they would contain the first 50% of all total samples and corresponding outputs.

Please let me know if this doesn't answer your question!

**Comment by walkeranthonyp** [comment]
> The variables that need to be passed through between restarts are listed here:
> 
> * original`j` (discussed above)
> * `n_CR`
> * `adapt_pCR`
> * `CR_burnin`
> * `CR`
> * `p_CR`
> * `del`
> * `L`
> * `t`
> * `m`

Also, can you double check this list of variables that need to be passed from the restart. I assume `chains` also needs to be passed, are there others?

**Comment by abbeyjohnson1** [comment]
I double checked them! Yes, `chains` needs to be passed, but not in the same sense as the crossover variables. This was a clarification I should have made sooner, my apologies.

Here's an attempt at describing my thought process: Variables like `chains` are wrapper parameters initialized by the user in the call script. The way that I envision the restart process working is the user creating a call script, running the initial simulation, and once that first initial run is completed, simply switching the `restart` wrapper parameter to `true` and manually running restarts without changing any other MCMC parameters. So, all the relevant parameters that are in that `wpars` sublist are passed through in the sense that they are re-initialized with each restart run. So the information disconnect between successive runs is happening with parameters only in the `mcmc` list.

So just to clarify, these are the parameters in the `mcmc` list that need to be passed through:
- `CR`
- `p_CR`
- `del`
- `L`
- `t`
- `m`

(Side note: The "original" `j` value also needs to be passed through but it's not a parameter in either of the lists.)

In my mind, it could be a bit inefficient to recode things for the parameters in the `wpars` sublist, since they are technically already being passed through in a sense. But please let me know if you want me to restructure this!

**Comment by walkeranthonyp** [comment]
OK, good. I have all of those. 

The way you describe the restart working makes total sense. But it assumes that everyone wanting to do a restart knows what they're doing! Which might not be the case. For example, someone could decide to run with more chains in the restart, which would lead to a code failure somewhere.

So to prevent this it's best to pass all the parameters for an MCMC run through from the original first run. Do we need all the parameters in the `wpars$mcmc` list? Or just a subset? Probably all right? It would be easy to do that.

**Comment by walkeranthonyp** [comment]
> * It could be informative to users if we carried over the "original j" value (i.e., total number of samples since the beginning of the MCMC process).

Assuming adapt_pCR is always TRUE, Isn't this `t` in the dynamic `mcmc` list?

**Comment by abbeyjohnson1** [comment]
That's a really good point - passing everything through from the original run to best avoid "user error." Yes, we would definitely want to pass through everything in the `wpars$mcmc` sublist.

**Comment by abbeyjohnson1** [comment]
Yes, as long as `adapt_pCR` is `true`, then `t` could be counter for the total number of samples. It was originally designed to be a separate counter in the crossover sub-algorithm, but I don't see any reason why this can't be changed to make it the `original j` as well. It will need a little bit of code edits though; for example, it will be re-initialized to 1 each time the `init_mcmc_dream` is called with each restart.

**Comment by walkeranthonyp** [comment]
Ah yeah, good catch. Assuming that if a restart is still in CR burnin mode then `t` does not want to be initialised at 1 right?

**Comment by abbeyjohnson1** [comment]
Yes, that's correct! It would need to be the total number of samples in the entire cumulative MCMC process. This is different from the `j` iteration value, which doesn't count the samples that are being "discarded" during burn-in. The `t` counter would need to count the samples that are being discarded.

**Comment by walkeranthonyp** [comment]
What about `d_star` - does that need passing through to a restart?

**Comment by walkeranthonyp** [comment]
Also, I just pushed some new edits. Please take a look at changes to `init_mcmc_dream` on lines 77 to 139 of `wrapper_functions_mcmc.R` and let me know if you see any issues with those changes.

**Comment by abbeyjohnson1** [comment]
No, `d_star` doesn't need to be passed through. It's a scalar value that's re-calculated with each proposal generation that's equivalent to the number of dimensions being updated.

**Comment by walkeranthonyp** [comment]
Alright, so the code that's online currently should have dealt with all the CR bugs issues in the original issue statement. 

`t` is the original j and represents the iteration of the full MCMC from initiation. 

I also changed `wpars$mcmc_pars` to `wpars$mcmc` and updated all the instances of this in the code (there were some incomplete updates to the mcmc_pars list).   

I haven't dealt with the output file issues yet. Nor the convergence testing or switch. Do we want to use the same number of iterations until convergence testing begins as we do for CR_burnin? Or should we keep them separate? I'll take a look at these issues once you've had a chance to look over my most recent code edits and have given any feedback.

**Comment by abbeyjohnson1** [comment]
Awesome!!! I'll test it out! Thanks for catching those variable names too! 

Re the iterations: In Vrugt's code, those are two different variables -- one for CR_burnin and a different one for the convergence testing switch. There's no mathematical reason for them to be the same numbers or different. But for the sake of generality and comprehensiveness, it might nice if they could be two different values. 

Sounds like a plan! I'll keep you updated!

**Comment by walkeranthonyp** [comment]
With all the updates in v1.3 all of these issues should now be fixed. All the updates in `dev_mcmc_bells` have been bugfixed and the MCMC code appears to be working well with the 3-mode mixture model test case. Re-reading some of the comments here it's possible that I messed up the use of `t` in certain calculations, I don't think so, but it's possible. This variable I relabelled and is now called `j_true`. I will check.
